zmq.hpp

```diff
@@ -152,6 +152,39 @@ typedef struct
 
 namespace zmq
 {
+
+#ifdef ZMQ_CPP11
+namespace detail
+{
+template<class T> using void_t = void;
+
+template<class Iter>
+using iter_value_t = typename std::iterator_traits<Iter>::value_type;
+
+template<class Range>
+using range_iter_t = decltype(
+  std::begin(std::declval<typename std::remove_reference<Range>::type &>()));
+
+template<class Range>
+using range_value_t = iter_value_t<range_iter_t<Range>>;
+
+template<class T, class = void> struct is_range : std::false_type
+{
+};
+
+template<class T>
+struct is_range<
+  T,
+  void_t<decltype(
+    std::begin(std::declval<typename std::remove_reference<T>::type &>())
```

Comment:
Isn't this too restrictive? std::begin isn't necessarily defined for every range, there might be a begin in a different namespace available via ADL. This also applies to the use in the meaaage_t ctor below.

Reply:
Fixed. I expected it to be harder to implement, but turned out to be quite simple.